# DSCI 100 Project: Final Report

**Authors**: 

**Predicting Usage of a Video Game Research Server**
A research group in the Department of Computer Science at UBC (PLAI) has a goal to aid in advancing the limits of artificial intelligence. They host a Minecraft server that records players' actions as they navigate through the world. Since running the project is so complex, they want to recruit players who will contribute large amounts of gameplay data, as well as make sure they have the resources to handle the amount of players they attract. In an attempt to optimize their players' contributions, they want to understand what demographics of players will have the greatest gameplay time. This project answers their question of how player demographics relate to total gameplay time. 

## The Question
**Question 2**: Which kinds of players are most likely to contribute a large amount of data, and how can we identify them for targeted recruitment?

**Predictive Question** Can we predict the total `played_hours` from the `players.csv` dataset based on `experience`.

**Response variable**:

`played_hours` - total play time and data contribution

**Explanatory variables**:

`experience` - player experience level

`subscribe` -  subscription status of player

The dataset allows us to examine whether `experience` or `subscribe` relates to total playtime. If certain experience levels or subscribtion status tend to play more, those groups may contribute more gameplay data. This helps identify the player demographics most likely to provide substantial data for the research team.


## # Import libraries
import pandas as pd
import altair as alt

# import the K-NN regression model
from sklearn.model_selection import GridSearchCV, cross_validate, train_test_split
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn import set_config

# import the K-NN regression model
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_errorData Description: The Players
The dataset used in this analysis is a player demographics and engagement dataset that contains 196 observations (rows) and 9 variables (columns). The dataset records information about individual players, their gaming experience level, subscription status, demographic details, and the amount of time they have spent playing.

- Number of observations: 196 (each row represents one unique player)
- Number of variables: 9
- Observational unit: Player-level data
- Purpose: Identify which demographic characteristics correspond to higher gameplay time (and therefore more data contributed)
  
**Variables**
- `played_hours` (numeric) - Total hours each player spent on the server
- `experience` (categorical) -  Self-reported experience level in Minecraft
- `gender` (categorical) - Player’s gender
- `subscribe` (Boolean) - Subscription status of player
- `hashedEmail` (String) - The players email, hashed for privacy
- `name` (String) - Name of player
- `age` (Numeric) - Age of player
- `individualID` (String) - Unique ID of player
- `organizationName` (String) - Organization name

**Issues to consider**
- Missing identifiers (`individualId`, `organizationName`).
- `hashedEmail` is anonymous for privacy.
- `played_hours` may include idle time, which could overestimate actual engagement.

Source of the original dataset:
https://drive.google.com/file/d/1Mw9vW0hjTJwRWx0bDXiSpYsO3gKogaPz/edit


In [1]:
# Import libraries
import pandas as pd
import altair as alt

# import the K-NN regression model
from sklearn.model_selection import GridSearchCV, cross_validate, train_test_split
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn import set_config

# import the K-NN regression model
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error

# Read the datasets
url= "https://raw.githubusercontent.com/sydlpeters/dsci-group-2025w1-group-101-1/refs/heads/main/data/players.csv"
# Load the dataset 
players = pd.read_csv(url)
players

experience  subscribe                                        hashedEmail  \
0          Pro       True  f6daba428a5e19a3d47574858c13550499be23603422e6...   
1      Veteran       True  f3c813577c458ba0dfef80996f8f32c93b6e8af1fa9397...   
2      Veteran      False  b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3...   
3      Amateur       True  23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4f...   
4      Regular       True  7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb...   
..         ...        ...                                                ...   
191    Amateur       True  b6e9e593b9ec51c5e335457341c324c34a2239531e1890...   
192    Veteran      False  71453e425f07d10da4fa2b349c83e73ccdf0fb3312f778...   
193    Amateur      False  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
194    Amateur      False  f19e136ddde68f365afc860c725ccff54307dedd13968e...   
195        Pro       True  d9473710057f7d42f36570f0be83817a4eea614029ff90...   

     played_hours       name             gender  age  individualId  \
0            30.3     Morgan               Male    9           NaN   
1             3.8  Christian               Male   17           NaN   
2             0.0      Blake               Male   17           NaN   
3             0.7      Flora             Female   21           NaN   
4             0.1      Kylie               Male   21           NaN   
..            ...        ...                ...  ...           ...   
191           0.0     Bailey             Female   17           NaN   
192           0.3     Pascal               Male   22           NaN   
193           0.0      Dylan  Prefer not to say   17           NaN   
194           2.3     Harlow               Male   17           NaN   
195           0.2      Ahmed              Other   91           NaN   

     organizationName  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
..                ...  
191               NaN  
192               NaN  
193               NaN  
194               NaN  
195               NaN  

[196 rows x 9 columns]

In [100]:
# Drop columns that are entirely missing(NaN)
players_tidy = players.drop(columns=["individualId", "organizationName", "name", "gender", "hashedEmail", "age","subscribe"])
# Preview tidy players_tidy
players_tidy.head()

experience  played_hours
0        Pro          30.3
1    Veteran           3.8
2    Veteran           0.0
3    Amateur           0.7
4    Regular           0.1

In [101]:
players_tidy.describe()

played_hours
count    196.000000
mean       5.845918
std       28.357343
min        0.000000
25%        0.000000
50%        0.100000
75%        0.600000
max      223.100000

**Summary Statistics**

Most players recorded very little playtime, with three-quarters spending under an hour in total (Q3 = 0.6 hours). Only a small number logged substantially more time, which raises the overall average (mean = 5.85 hours; median = 0.1 hours) and affects the dataset (STD = 28.36 hours). **Most players played very little (close to 0 hours), but a few played way more, making the dataset highly uneven.**

In [102]:
# Experienced comapred with Average played hours played

# Summary statistics and graphs should only be done on training data

plot1= alt.Chart(players, title="Fig 1. Total Hours Played per Experience Group").mark_bar().encode(
    x=alt.X("experience:N").title("Experience Level"),
    y=alt.Y("mean(played_hours):Q").title("Average hours Played(hr)"),
    color = alt.Color("experience:N").title("Experience Level")
)
plot1


alt.Chart(...)

In [103]:
# Experienced comapred with Average played hours played
plot2= alt.Chart(players, title="Fig. 2 Average Hours Played per Experience Group").mark_bar().encode(
    x=alt.X("experience:N").title("Experience Level"),
    y=alt.Y("played_hours:Q").title("Average hours Played(hr)"),
    color = alt.Color("experience:N").title("Experience Level")
)
plot2

alt.Chart(...)

In [104]:
# Subscription status comapred with Average hours played
plot3= alt.Chart(players, title="Fig. 3 Average Hours by Subscribed/Unsubscribed").mark_bar().encode(
    x=alt.X("subscribe:N").title("Subscription Status"),
    y=alt.Y("mean(played_hours)").title("Average Hours Played(hr)"),
    color = alt.Color("subscribe:N").title("Subscription Status")
)
plot3

alt.Chart(...)

In [105]:

plot4= alt.Chart(players, title="Fig. 4 Average Hours Played by Gender").mark_bar().encode(
    x=alt.X("gender:N").title("Gender"),
    y=alt.Y("mean(played_hours)").title("Average Hours Played(hr)"),
    color = alt.Color("gender:N", legend=None).title("Gender")
)
plot4

alt.Chart(...)

In [106]:

plot5= alt.Chart(players, title="Fig. 5 Average Hours Played by each Age").mark_circle(opacity=0.8).encode(
    x=alt.X("age:N", scale=alt.Scale(domain=[i for i in range(2, 102, 2)])).title("Age"),
    y=alt.Y("mean(played_hours)").title("Average Hours Played(hr)"),
    color = alt.Color("age:N", legend=None).title("Age")
)
plot5

alt.Chart(...)

In [107]:
experience_map = {
    "Beginner" : 0,
    "Regular": 1,
    "Amateur": 2,
    "Veteran": 3,
    "Pro": 4
}
# Put above
players_tidy["experience"] = players_tidy["experience"].map(experience_map)
players_tidy

experience  played_hours
0             4          30.3
1             3           3.8
2             3           0.0
3             2           0.7
4             1           0.1
..          ...           ...
191           2           0.0
192           3           0.3
193           2           0.0
194           2           2.3
195           4           0.2

[196 rows x 2 columns]

In [108]:
# split data!
players_training, players_testing = train_test_split(
    players_tidy, test_size=0.2, random_state=1234
)

# set target and predictors
players_x_train = players_training[["experience"]]
players_y_train = players_training["played_hours"]

players_x_test = players_testing[["experience"]]
players_y_test = players_testing["played_hours"]

players_training

experience  played_hours
114           0           1.0
187           2           0.0
39            2           0.0
25            1           0.6
131           2           0.0
..          ...           ...
152           0           0.2
116           3           0.0
53            2           0.2
38            3           0.0
47            1           0.0

[156 rows x 2 columns]

In [109]:
# Experienced comapred with Average played hours played
plot6 = alt.Chart(players_training, title="Fig. 6 Average Hours Played per Experience Group on training set").mark_bar().encode(
    x=alt.X("experience:N").title("Experience Level"),
    y=alt.Y("played_hours:Q").title("Hours Played(hr)"),
    color = alt.Color("experience:N").title("Experience Level")
)
plot6

alt.Chart(...)

In [110]:
# preprocess the data, make the pipeline
players_preprocessor = make_column_transformer(
    (
        StandardScaler(),
        [
            "experience",
            ],
    )
)
players_pipeline = make_pipeline(players_preprocessor, KNeighborsRegressor())

players_cv = pd.DataFrame(
    cross_validate(
        players_pipeline,
        players_x_train,
        players_y_train,
        cv=5,
        scoring="neg_root_mean_squared_error",
        return_train_score=True,
    )
)
players_cv

fit_time  score_time  test_score  train_score
0  0.004266    0.002295  -19.160247   -27.106263
1  0.003216    0.002147   -5.803200   -21.794819
2  0.003107    0.001941  -29.223031   -16.374080
3  0.003083    0.001929   -3.798582   -21.907992
4  0.003085    0.001921  -31.885128   -15.327369

In [111]:
# create the 5-fold GridSearchCV object
param_grid = {
    "kneighborsregressor__n_neighbors": range(1, 50, 1),
}

players_tuned = GridSearchCV(
    estimator=players_pipeline,
    param_grid=param_grid,
    cv=5,
    scoring="neg_root_mean_squared_error",
)

# fit the GridSearchCV object
players_result = pd.DataFrame(players_tuned.fit(players_x_train, players_y_train).cv_results_)

players_result

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.003697      0.000778         0.002243        0.000578   
1        0.003110      0.000084         0.001947        0.000053   
2        0.003247      0.000234         0.002005        0.000134   
3        0.003050      0.000052         0.001928        0.000080   
4        0.003008      0.000035         0.001925        0.000068   
5        0.003000      0.000008         0.001883        0.000006   
6        0.003001      0.000020         0.001926        0.000088   
7        0.003042      0.000094         0.001912        0.000048   
8        0.002995      0.000008         0.001915        0.000065   
9        0.002987      0.000015         0.001913        0.000064   
10       0.002982      0.000021         0.001889        0.000036   
11       0.002975      0.000003         0.001910        0.000062   
12       0.003089      0.000225         0.001898        0.000070   
13       0.003193      0.000415         0.001888        0.000025   
14       0.003032      0.000062         0.001933        0.000063   
15       0.003021      0.000052         0.001877        0.000010   
16       0.002976      0.000007         0.001912        0.000081   
17       0.002984      0.000034         0.001922        0.000075   
18       0.002979      0.000006         0.001893        0.000023   
19       0.002979      0.000006         0.001922        0.000092   
20       0.002976      0.000007         0.001911        0.000067   
21       0.002991      0.000042         0.001899        0.000020   
22       0.002980      0.000017         0.001920        0.000060   
23       0.002975      0.000021         0.001879        0.000007   
24       0.002976      0.000036         0.001905        0.000068   
25       0.002991      0.000046         0.001927        0.000060   
26       0.005210      0.004202         0.001932        0.000038   
27       0.003144      0.000302         0.001937        0.000064   
28       0.002990      0.000035         0.001928        0.000067   
29       0.008317      0.010673         0.002055        0.000277   
30       0.003027      0.000058         0.001955        0.000061   
31       0.002967      0.000008         0.001926        0.000032   
32       0.003323      0.000703         0.001945        0.000070   
33       0.002970      0.000018         0.001933        0.000070   
34       0.003100      0.000136         0.001931        0.000029   
35       0.003381      0.000556         0.002105        0.000197   
36       0.003002      0.000008         0.001970        0.000066   
37       0.003010      0.000034         0.002128        0.000359   
38       0.003004      0.000021         0.001987        0.000118   
39       0.002972      0.000010         0.001917        0.000008   
40       0.002970      0.000008         0.001954        0.000067   
41       0.002964      0.000007         0.001950        0.000066   
42       0.002974      0.000019         0.001918        0.000007   
43       0.002982      0.000005         0.001962        0.000063   
44       0.003023      0.000097         0.001963        0.000065   
45       0.002997      0.000029         0.001936        0.000019   
46       0.003012      0.000035         0.002173        0.000395   
47       0.002981      0.000017         0.001927        0.000023   
48       0.002971      0.000016         0.001957        0.000065   

    param_kneighborsregressor__n_neighbors  \
0                                        1   
1                                        2   
2                                        3   
3                                        4   
4                                        5   
5                                        6   
6                                        7   
7                                        8   
8                                        9   
9                                       10   
10                                      11   
11                                      12   
12   

In [112]:
# Add Visualization to compare the ks
players_result["rmse"] = -players_result["mean_test_score"]
players_k_plot= alt.Chart(players_result).mark_line().encode(
    x=alt.X("param_kneighborsregressor__n_neighbors:Q", title="k (Number of Neighbours)"),
    y=alt.Y("rmse:Q", title="5-fold CV RMSE")
)
players_k_plot

alt.Chart(...)

In [113]:
# Retrieve the CV scores

# get the best parameter values
players_min = players_tuned.best_params_
players_min

{'kneighborsregressor__n_neighbors': 39}

In [114]:
players_best_RMSPE = -players_tuned.best_score_
players_best_RMSPE

np.float64(16.028531937252072)

In [115]:
knn = KNeighborsRegressor(n_neighbors = 39)
player_pipe = make_pipeline(players_preprocessor, knn)
player_pipe.fit(players_x_train, players_y_train)

players_y_predictions = player_pipe.predict(players_x_train)

players_predictions = players_training.assign(prediction =players_train_predictions)
players_predictions

experience  played_hours  prediction
114           0           1.0    1.410256
187           2           0.0    7.261538
39            2           0.0    7.261538
25            1           0.6    6.494872
131           2           0.0    7.261538
..          ...           ...         ...
152           0           0.2    1.410256
116           3           0.0    0.464103
53            2           0.2    7.261538
38            3           0.0    0.464103
47            1           0.0    6.494872

[156 rows x 3 columns]

In [116]:
rmse_df = pd.DataFrame({
    "actual": players_y_train, 
    "predicted": players_train_predictions
})
train_rmse = mean_squared_error(players_y_train, players_train_predictions) ** 0.5
train_rmse

np.float64(19.51054219122817)

In [117]:

players_plot = (alt.Chart(players_predictions).mark_circle(opacity=0.7).encode(
    x=alt.X("experience").title("Experience").scale(zero=False),
    y=alt.Y("played_hours").title("played_hours").scale(zero=False)
)+
alt.Chart(
    players_predictions,
    title= "K=39"
).mark_line(
    color="black"
).encode(
    x="experience",
    y="prediction"
))
players_plot


alt.LayerChart(...)

In [118]:

# create the 5-fold GridSearchCV object
param_grid = {
    "kneighborsregressor__n_neighbors": range(1, 50, 1),
}

players_tuned = GridSearchCV(
    estimator=players_pipeline,
    param_grid=param_grid,
    cv=5,
    scoring="neg_root_mean_squared_error",
)


players_result = pd.DataFrame((players_tuned.fit(players_x_train, players_y_train)).cv_results_)

players_result

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.003385      0.000206         0.002379        0.000575   
1        0.003111      0.000025         0.001942        0.000010   
2        0.003100      0.000084         0.002862        0.001855   
3        0.003025      0.000007         0.001933        0.000071   
4        0.003011      0.000011         0.001912        0.000031   
5        0.003011      0.000012         0.001937        0.000082   
6        0.003296      0.000620         0.001887        0.000019   
7        0.003009      0.000018         0.001980        0.000180   
8        0.003003      0.000030         0.001913        0.000064   
9        0.002984      0.000005         0.001878        0.000006   
10       0.002975      0.000006         0.002013        0.000216   
11       0.002985      0.000022         0.001922        0.000080   
12       0.003017      0.000054         0.001894        0.000024   
13       0.002991      0.000017         0.001907        0.000070   
14       0.003223      0.000489         0.001881        0.000007   
15       0.003352      0.000560         0.002027        0.000169   
16       0.003097      0.000049         0.002147        0.000318   
17       0.003033      0.000009         0.001919        0.000011   
18       0.003023      0.000014         0.002092        0.000294   
19       0.003031      0.000048         0.001953        0.000061   
20       0.003169      0.000333         0.001907        0.000006   
21       0.002996      0.000009         0.001939        0.000068   
22       0.003012      0.000036         0.001902        0.000012   
23       0.003107      0.000221         0.001939        0.000068   
24       0.003000      0.000011         0.001942        0.000071   
25       0.003007      0.000025         0.001930        0.000036   
26       0.002991      0.000011         0.001948        0.000090   
27       0.003024      0.000042         0.001989        0.000064   
28       0.003003      0.000023         0.001911        0.000008   
29       0.002991      0.000014         0.001939        0.000064   
30       0.003011      0.000031         0.001915        0.000006   
31       0.002988      0.000006         0.001948        0.000068   
32       0.002982      0.000005         0.001939        0.000066   
33       0.002990      0.000011         0.001907        0.000009   
34       0.002988      0.000007         0.001949        0.000078   
35       0.003073      0.000177         0.001951        0.000060   
36       0.003508      0.000841         0.002062        0.000155   
37       0.003039      0.000019         0.002002        0.000097   
38       0.003019      0.000009         0.001941        0.000004   
39       0.003018      0.000020         0.001973        0.000073   
40       0.002996      0.000004         0.001967        0.000071   
41       0.002992      0.000004         0.001939        0.000008   
42       0.002992      0.000007         0.001972        0.000061   
43       0.002986      0.000008         0.001975        0.000068   
44       0.003171      0.000333         0.001939        0.000008   
45       0.002998      0.000024         0.001984        0.000109   
46       0.002977      0.000010         0.001930        0.000010   
47       0.002975      0.000011         0.001989        0.000060   
48       0.002975      0.000010         0.001970        0.000075   

    param_kneighborsregressor__n_neighbors  \
0                                        1   
1                                        2   
2                                        3   
3                                        4   
4                                        5   
5                                        6   
6                                        7   
7                                        8   
8                                        9   
9                                       10   
10                                      11   
11                                      12   
12   

In [119]:
players_min = players_tuned.best_params_
players_min

{'kneighborsregressor__n_neighbors': 39}

In [120]:
players_best_RMSPE = -players_tuned.best_score_
players_best_RMSPE

np.float64(16.028531937252072)

In [121]:
players_tidy

experience  played_hours
0             4          30.3
1             3           3.8
2             3           0.0
3             2           0.7
4             1           0.1
..          ...           ...
191           2           0.0
192           3           0.3
193           2           0.0
194           2           2.3
195           4           0.2

[196 rows x 2 columns]

In [122]:
best_model = players_tuned.best_estimator_
all_pred = best_model.predict(players_tidy[["experience"]])
players_tidy_predictions =players_tidy.assign(prediction=all_pred)
players_tidy_predictions

experience  played_hours  prediction
0             4          30.3    1.305128
1             3           3.8    0.464103
2             3           0.0    0.464103
3             2           0.7    7.261538
4             1           0.1    6.494872
..          ...           ...         ...
191           2           0.0    7.261538
192           3           0.3    0.464103
193           2           0.0    7.261538
194           2           2.3    7.261538
195           4           0.2    1.305128

[196 rows x 3 columns]

In [125]:
all_rmse = mean_squared_error(
    players_tidy_predictions["played_hours"],
    players_tidy_predictions["prediction"]
) ** 0.5

all_rmse

np.float64(28.04716413446685)

In [128]:

players_plot = (alt.Chart(players_tidy_predictions).mark_circle(opacity=0.7).encode(
    x=alt.X("experience").title("Experience").scale(zero=False),
    y=alt.Y("played_hours").title("played_hours").scale(zero=False)
)+
alt.Chart(
    players_predictions,
    title= "K=39"
).mark_line(
    color="black"
).encode(
    x="experience",
    y="prediction"
))
players_plot

alt.LayerChart(...)